src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -1991,13 +1993,105 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
         }
 
         /// <summary>
-        /// Gives the preview as a list of NuGetProjectActions that will be performed to uninstall
+        /// Gives the preview as a list of NuGetProjectActions that will be performed to uninstall for many NuGetProjects.
+        /// </summary>
+        /// <param name="nuGetProjects"></param>
+        /// <param name="packageId"></param>
+        /// <param name="uninstallationContext"></param>
+        /// <param name="nuGetProjectContext"></param>
+        /// <param name="token"></param>
+        /// <returns></returns>
+        public async Task<IEnumerable<(NuGetProject, NuGetProjectAction)>> PreviewProjectsUninstallPackageAsync(
+            IReadOnlyCollection<NuGetProject> nuGetProjects,
+            string packageId,
+            UninstallationContext uninstallationContext,
+            INuGetProjectContext nuGetProjectContext,
+            CancellationToken token)
+        {
+            if (nuGetProjects == null)
+            {
+                throw new ArgumentNullException(nameof(nuGetProjects));
+            }
+
+            if (packageId == null)
+            {
+                throw new ArgumentNullException(nameof(packageId));
+            }
+
+            if (uninstallationContext == null)
+            {
+                throw new ArgumentNullException(nameof(uninstallationContext));
+            }
+
+            if (nuGetProjectContext == null)
+            {
+                throw new ArgumentNullException(nameof(nuGetProjectContext));
+            }
+
+            var uninstallActionsForProjects = new ConcurrentBag<(NuGetProject, NuGetProjectAction)>();
+
+            var options = new ExecutionDataflowBlockOptions()
+            {
+                MaxDegreeOfParallelism = Environment.ProcessorCount
+            };
+            var actionBlock = new ActionBlock<ProjectUninstallActionSpec>(GetProjectNuGetUninstallProjectActionsAsync, options);
+
+            foreach (var nuGetProject in nuGetProjects)
+            {
+                await actionBlock.SendAsync(new ProjectUninstallActionSpec(
+                                                nuGetProject,
+                                                packageId,
+                                                uninstallationContext,
+                                                nuGetProjectContext,
+                                                uninstallActionsForProjects,
+                                                token)
+                                            );
+            }
+
+            actionBlock.Complete();
+            await actionBlock.Completion;
+
+            return uninstallActionsForProjects;
+        }
+
+        private async Task GetProjectNuGetUninstallProjectActionsAsync(ProjectUninstallActionSpec projectUninstallActionSpec)
```

Comment:
I'm not sure this is parallelizing the restore runs as we want them to be. 

This calls 

```cs
        private async Task<IEnumerable<NuGetProjectAction>> PreviewUninstallPackageInternalAsync(NuGetProject nuGetProject, Packaging.PackageReference packageReference,
```

which calls:

```cs
        public async Task<BuildIntegratedProjectAction> PreviewBuildIntegratedProjectActionsAsync(
            BuildIntegratedNuGetProject buildIntegratedProject,
            IEnumerable<NuGetProjectAction> nuGetProjectActions,
            INuGetProjectContext nuGetProjectContext,
            CancellationToken token)
```
which runs restore.

Reply:
I believe it's ok, because we're already doing it for `install action`.
Here `PreviewBuildIntegratedProjectActionsAsync`   [Line 2651](https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697) is just simple façade for https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697-L2702

Here `PreviewBuildIntegratedProjectActionsAsync` on [Line 2697](https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697) is already called for `install action` from https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1601

Here below is how call stack looks like until  we reach above [NuGetPackageManager.cs#L1601](https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1601) for install action: 
1. If it's 'install action'?
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L776

2. Then call this `GetInstallActionsAsync` method:
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs#L780

3. `GetInstallActionsAsync` method is here:
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs#L256

4. From inside here call method `PreviewProjectsInstallPackageAsync`:
https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetProjectManagerService.cs#L294

5.  We call method `PreviewBuildIntegratedProjectsActionsAsync` : https://github.com/NuGet/NuGet.Client/blob/da3133bd51bd5db4d4599873770d73c4a69028f5/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L1598-L1601

Reply:
In 4, preview Project*s*, plural, is calling Project, singular. 

In the current implementation you are calling restore individually. If I'm following the code correctly, uninstall will end up calling: https://github.com/NuGet/NuGet.Client/blob/087a7113ad70ae4d4c538eba0f387271af0c649b/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L141-L151 and not 
https://github.com/NuGet/NuGet.Client/blob/087a7113ad70ae4d4c538eba0f387271af0c649b/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L194-L204.

Reply:
@nkolev92 
Uninstall action is calling correct correct dgSpec method (plural one).
Here call stack starting parallel job call:
1. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2057

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2070-L2075

2. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2169

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2186-L2188

3. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2651

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2678-L2688

4. https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697

From here call this one:
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2858-L2870

5. Here correct plural dgSpec method.
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs#L194

Reply:
That's because:

PreviewBuildIntegratedProjectActionsAsync calls PreviewBuildIntegratedProjectsActionsAsync

Basically from: 
https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2169

to

https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2651

to 

https://github.com/NuGet/NuGet.Client/blob/78d1d9eb03acf9efcdea76044e8e5e4662dd2e22/src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs#L2697

If you put a break point in DependencyGraphRestoreUtility.cs.PreviewRestoreProjectsAsync, you will see that it's called twice, while it should be called once.

Reply:
@nkolev92  
Ok, done. 
Please check. I see another few seconds gain on large project like OrchardCore. 😄